### ISSUE TITLE
[Feature Request] File Share Progress Bar

### ISSUE THREAD

**Issue by busterbeam** [andyholmes/gnome-shell-extension-gsconnect#1511]
https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/6a7bef6b61d419bad90665e27487484df8511d4f/src/service/daemon.js#L533-L552

I'm not accustomed to DBus is this where it calls the service to send file (providing it has the send file arguments/parameters)?  Scanning through the [DBusProxy](https://docs.gtk.org/gio/class.DBusProxy.html) documentation I'm finding it hard to understand where you can pass file data in chunks so you can measure the rate of progress.

Even using bluetooth file transfer do I get a progress bar in GNOME so why not for the same faster WiFi connection

**Comment by andyholmes** [comment]
If you want to add a progress bar over D-Bus, you'll have to modify the [`Core.Transfer`](https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/main/src/service/core.js#L479) class and then modify the [`GSConnectSharePlugin.shareFile()`](https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/main/src/service/plugins/share.js#L262) method to take advantage of it. Currently GSConnect doesn't export any information about the progress of transfers.

**Comment by ferdnyc** [comment]
You might even need to go down into [`Core.Channel`](https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/main/src/service/core.js#L121) which defines some of the methods for wrangling `Core.Packet`s.

The `this.channel.upload()` called in `Core.Transfer` is implemented in [`Lan.Channel`](https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/main/src/service/backends/lan.js#L879), but it's not immediately clear to me where best to monitor transfer progress.

(Unrelated to this request) there's a Gio definition for [`FileProgressCallback`](https://docs.gtk.org/gio/callback.FileProgressCallback.html) that will be updated by potentially-time-consuming operations like `Gio.File.copy()` and `Gio.File.move()`, that's how the file managers show progress on file operations.

Unfortunately the IOStream classes don't seem to have built-in support for the same sort of reporting (probably because a stream has no fixed length). That makes the [`Gio.OutputStream.splice_async`](https://docs.gtk.org/gio/method.OutputStream.splice_async.html) and `Gio.Inputstream.read_all_async()`](https://docs.gtk.org/gio/method.InputStream.read_all_async.html) methods we use to transfer data over the stream a bit trickier to monitor. Their `GAsyncReadyCallback`s only fire on completion of the operation.

It may be possible to break stream transfers up so they're done in multiple chunks in a loop, then progress could be updated each time through. But I don't _think_ Gio will generate any progress data itself, at least not any that calling code can get at.

**Comment by andyholmes** [comment]
Yeah, it's been awhile since I looked into that part of GSConnect. I'd expect this to take more or less three steps:

1. `Core.Transfer`: implement progress tracking

    Instead of using `Gio.OutputStream.splice_async()`, you'd use e.g. [`Gio.InputStream.read_bytes_async()`](https://gjs-docs.gnome.org/gio20/gio.inputstream#method-read_bytes_async) and keep looping while tracking bytes read vs bytes expected. Either using a property or a signal to notify of changes.

2. `Core.Transfer`: implement D-Bus interface

    Each transfer would probably have to export itself as a simple D-Bus interface, otherwise you could write a simple transfer manager that tracked each transfer and export that as a single interface.

3. Elsewhere: implement D-Bus proxy/client

    Once transfers are being exported on D-Bus, you'll need to write a small `Gio.DBusProxy` subclass or similar to consume the information so it can be used to display the progress of transfers somewhere.

This isn't all terribly complicated, but it will take a bit of commitment to get it over the finish line :)

**Comment by busterbeam** [comment]
@andyholmes thank you for your input it was very helpful.  I will attempt to implement this

**Comment by busterbeam** [comment]
@andyholmes thank you for your input it was very helpful.  I will attempt to implement this

**Comment by busterbeam** [comment]
@andyholmes thank you for your input it was very helpful.  I will attempt to implement this
